spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java

```diff
@@ -91,112 +108,51 @@ public void setProducerListener(ProducerListener<K, V> producerListener) {
 	}
 
 	@Override
-	public Future<RecordMetadata> send(V data) {
+	public ListenableFuture<SendResult<K, V>> send(V data) {
 		return send(this.defaultTopic, data);
 	}
 
 	@Override
-	public Future<RecordMetadata> send(K key, V data) {
+	public ListenableFuture<SendResult<K, V>> send(K key, V data) {
 		return send(this.defaultTopic, key, data);
 	}
 
 	@Override
-	public Future<RecordMetadata> send(int partition, K key, V data) {
+	public ListenableFuture<SendResult<K, V>> send(int partition, K key, V data) {
 		return send(this.defaultTopic, partition, key, data);
 	}
 
 	@Override
-	public Future<RecordMetadata> send(String topic, V data) {
+	public ListenableFuture<SendResult<K, V>> send(String topic, V data) {
 		ProducerRecord<K, V> producerRecord = new ProducerRecord<>(topic, data);
 		return doSend(producerRecord);
 	}
 
 	@Override
-	public Future<RecordMetadata> send(String topic, K key, V data) {
+	public ListenableFuture<SendResult<K, V>> send(String topic, K key, V data) {
 		ProducerRecord<K, V> producerRecord = new ProducerRecord<>(topic, key, data);
 		return doSend(producerRecord);
 	}
 
 	@Override
-	public Future<RecordMetadata> send(String topic, int partition, V data) {
+	public ListenableFuture<SendResult<K, V>> send(String topic, int partition, V data) {
 		ProducerRecord<K, V> producerRecord = new ProducerRecord<K, V>(topic, partition, null, data);
 		return doSend(producerRecord);
 	}
 
 	@Override
-	public Future<RecordMetadata> send(String topic, int partition, K key, V data) {
+	public ListenableFuture<SendResult<K, V>> send(String topic, int partition, K key, V data) {
 		ProducerRecord<K, V> producerRecord = new ProducerRecord<>(topic, partition, key, data);
 		return doSend(producerRecord);
 	}
 
 	@Override
-	public Future<RecordMetadata> send(Message<?> message) {
+	public ListenableFuture<SendResult<K, V>> send(Message<?> message) {
 		ProducerRecord<K, V> producerRecord = messageToProducerRecord(message);
```

Comment:
Indeed @artembilan, I also think this is a perfect use case for `.map(this::messageToProducerRecord)`.

Please notice that for more heavy transformations that could involve some latency, something like `.flatMap(this::fooMethod)` with `fooMethod` returning `Mono` or `Flux` would be a better choice in order to avoid to block too much the processing. That's why we have created [Reactive codec in Spring Reactive](https://github.com/spring-projects/spring-reactive/tree/master/src/main/java/org/springframework/core/codec) that are not specific to a protocol (they could be used for HTTP, messaging, Websocket, file, etc.). Most serialization libraries are still blocking so we stick to process small objects thanks to [JsonObjectDecoder](https://github.com/spring-projects/spring-reactive/blob/master/src/main/java/org/springframework/core/codec/support/JsonObjectDecoder.java) for example, but some are already asynchronous like Aalto XML, see [@poutsma PR about that](https://github.com/spring-projects/spring-reactive/pull/77).